test/Microsoft.IdentityModel.Tokens.Tests/EventBasedLRUCacheTests.cs

```diff
@@ -0,0 +1,323 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.TestUtils;
+using Xunit;
+
+namespace Microsoft.IdentityModel.Tokens.Tests
+{
+    public class EventBasedLRUCacheTests
+    {
+        [Fact]
+        public void Contains()
+        {
+            TestUtilities.WriteHeader($"{this}.Contains");
+            var context = new CompareContext($"{this}.Contains");
+            using (var cache = new EventBasedLRUCache<int?, string>(10, TaskCreationOptions.LongRunning))
+            {
+                cache.SetValue(1, "one");
+                if (!cache.Contains(1))
+                    context.AddDiff("Cache should contain the key value pair {1, 'one'}, but the Contains() method returned false.");
+
+                cache.TryRemove(1, out _);
+                if (cache.Contains(1))
+                    context.AddDiff("The key value pair {1, 'one'} should have been removed from the cache, but the Contains() method returned true.");
+
+                try
+                {
+                    cache.Contains(null);
+                    context.AddDiff("The parameter passed into the Contains() method was null, but no exception was thrown.");
+                }
+                catch (Exception ex)
+                {
+                    if (!ex.GetType().Equals(typeof(ArgumentNullException)))
+                        context.AddDiff("The exception type thrown by Contains(null) was not of type ArgumentNullException.");
+                }
+
+                TestUtilities.AssertFailIfErrors(context);
+            }
+        }
+
+        [Fact]
+        public void RemoveExpiredValues()
+        {
+            TestUtilities.WriteHeader($"{this}.RemoveExpiredValues");
+            var context = new CompareContext($"{this}.RemoveExpiredValues");
+            using (var cache = new EventBasedLRUCache<int, string>(11, TaskCreationOptions.LongRunning))
+            {
+                for (int i = 0; i <= 10; i++)
+                {
+                    // Only even values should expire.
+                    if (i % 2 == 0)
+                        cache.SetValue(i, i.ToString(), DateTime.UtcNow + TimeSpan.FromSeconds(5));
+                    else
+                        cache.SetValue(i, i.ToString());
+                }
+
+                Thread.Sleep(5000);
```

Comment:
>Thread.Sleep(5000); [](start = 16, length = 19)

Any reason this is still needed?

Reply:
This Thread.Sleep() is specifically added so that values in the cache have time to expire.

Reply:
Hmm, perhaps a shorter expiry time then? can be 1ms or even 0ms, right? either it has an expiration or not? Not a fan of waiting 5s for any reason in a utest

---
In reply to: [590681840](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1605#discussion_r590681840) [](ancestors = 590681840)

Reply:
+1 

pass expiry as UtcNow then waiting any amount should be fine right?

Reply:
If the _removeExpiredValues flag is set to true, the set API will not add expired values. This can be an issue when when just using Utc.Now, as the item could expire before it's added.

Reply:
understood, but is it a real problem for the ut?